lms/static/scripts/frontend_apps/utils/onedrive-picker-client.js

```diff
@@ -0,0 +1,73 @@
+import { PickerCanceledError } from './google-picker-client';
+import { loadOneDriveAPI } from './onedrive-api-client';
+
+export class OneDrivePickerClient {
+  /**
+   * @param {object} options
+   * @param {string} options.clientId
+   * @param {string} options.redirectURI - the URL that is used to launch the picker
+   */
+  constructor({ clientId, redirectURI }) {
+    this._onLoadOneDriveAPI = loadOneDriveAPI();
+    this._clientId = clientId;
+    this._redirectURI = redirectURI;
+  }
+
+  /**
+   * Opens the OneDrive picker
+   *
+   * @returns Promise<{url: string}> - the URL of the file
+   * @throws {Error} - there are two types of errors:
+   *   - PickerCancelledError: raised when the user cancels the OneDrive dialog picker
+   *   - Error: raised when (1) the OneDrive client fails to load, or when (2)
+   *       the picker fails with internal or backend problem.
+   */
+  async showPicker() {
+    const oneDrive = await this._onLoadOneDriveAPI;
+    return new Promise((resolve, reject) => {
+      const success = (/** @type {any} */ file) => {
+        const sharingURL = file.value[0].permissions[0].link.webUrl;
```

Comment:
So far, this is what I have concluded:

* The OneDrive picker or REST API doesn't provide a built-in method to encode the URL
* The `permission` array returned by the picker doesn't fully conform with the `permission` returned by the REST API (https://docs.microsoft.com/en-us/graph/api/resources/permission?view=graph-rest-1.0)
* Because there is not documentation about the `file` structure returned as argument to the `success` callback , I can't guaranty `webUrl` be always available.  Under normal circumstances, `webUrl` seems to be there.

Reply:
> Because there is not documentation about the file structure returned as argument to the success callback 

I see documentation at the URL linked in the JSDoc comment for the class. https://docs.microsoft.com/en-us/onedrive/developer/controls/file-pickers/js-v72/open-file?view=odsp-graph-online says:

> When the user is done picking file(s), the success callback receives response object. This object contains properties, include value property which is a collection of Item resource with a subset of the item's properties.

Where _item resource_ is a link to https://docs.microsoft.com/en-us/onedrive/developer/rest-api/resources/driveitem?view=odsp-graph-online. That page's description of the `webUrl` property implies that it is always present as, unlike other properties, it does not mention conditions under which it is present.

Reply:
About the `permission` collection it states that:

> The set of permissions for the item. Read-only. Nullable.

So I understand it can be an `Array` or `null`.